emcc.py

```diff
@@ -1344,6 +1349,17 @@ def check(input_file):
             logger.warning('BINARYEN_ASYNC_COMPILATION requested, but disabled because of user options. ' + warning)
           elif 'BINARYEN_ASYNC_COMPILATION=0' not in settings_changes:
             logger.warning('BINARYEN_ASYNC_COMPILATION disabled due to user options. ' + warning)
+        # Swappable wasm module/asynchronous wasm compilation requires an indirect stub
+        # function generated to each function export from wasm module, so cannot use the
+        # concise form of grabbing exports that does not need to refer to each export individually.
+        if shared.Settings.SWAPPABLE_ASM_MODULE == 1 and not shared.Settings.DECLARE_ASM_MODULE_EXPORTS:
+          shared.Settings.DECLARE_ASM_MODULE_EXPORTS = 1
+          logger.warning('Enabling -s DECLARE_ASM_MODULE_EXPORTS=1 since -s SWAPPABLE_ASM_MODULE=1 is used')
```

Comment:
the this and the modularize handler could be put close together, behind an `if not shared.Settings.DECLARE_ASM_MODULE_EXPORTS`. I also think it's simpler to error on them instead of warning (warnings might be missed).

Reply:
If user is specifying conditions where the option `DECLARE_ASM_MODULE_EXPORTS=0` is not possible to be used (e.g. `-O3`), it is not the end of the world to have `DECLARE_ASM_MODULE_EXPORTS` option revert to `=1` state. Code will build, but with an omitted optimization.

In build systems it gets really uncomfortable to have to know about combinations of different flags that are mutually incompatible, because they cannot then be specified orthogonally. (`--separate-asm` and `-s WASM=1` are currently an example of this, would be better for `--separate-asm` to just silently get ignored when `-s WASM=1` is in effect)

Ideally the limitation that MetaDCE is not compatible with `DECLARE_ASM_MODULE_EXPORTS=0` could be lifted in the future (and preamble refactored so that `SWAPPABLE_ASM_MODULE` option would be deprecated as unnecessary? though that's probably another conversation altogether), so users would not need to populate their build systems with special case code of "if -O3 or -Oz and we'd like DECLARE_ASM_MODULE_EXPORTS=0, don't actually specify DECLARE_ASM_MODULE_EXPORTS=0 (TODO: perhaps specify it in the future)". When/if later MetaDCE is improved to support DECLARE_ASM_MODULE_EXPORTS=0, then it will be a backwards breaking change in users' build systems when they combine `DECLARE_ASM_MODULE_EXPORTS=0` with `-O3`, older compilers would start aborting the builds if emcc.py early-aborted here, whereas it would be reasonable for `DECLARE_ASM_MODULE_EXPORTS=0` to just have gotten ignored.

I'd prefer to keep these as warnings if that's ok?

Reply:
Ok.